content/courses/advanced-ada/chapters/access_types.rst

```diff
@@ -294,14 +294,402 @@ Unchecked Access
+       I_2 := I_1;
+
+       Put_Line ("I_1 = " & Integer'Image (I_1.all));
+       Put_Line ("I_2 = " & Integer'Image (I_2.all));
+
+       Put_Line ("Freeing I_1");
+       Free (I_1);
+
+       if I_1 /= null then
+          Put_Line ("I_1 = " & Integer'Image (I_1.all));
+       end if;
+
+       if I_2 /= null then
+          Put_Line ("I_2 = " & Integer'Image (I_2.all));
+       end if;
+    end Show_Unchecked_Deallocation;
+
+In this example, we allocate an object for :ada:`I_1` and make :ada:`I_2` point
+to the same object. Note, however, that we call :ada:`Free` for :ada:`I_1`, but
+not for :ada:`I_2`. Therefore, after that call to :ada:`Free`,
+:ada:`I_1 = null`, but :ada:`I_2` still points to the object originally
+allocated for :ada:`I_1`.
+
+As mentioned previously, when deallocating an object via a call to :ada:`Free`,
+the corresponding access value is set to :ada:`null`. In this case, we can
+check for :ada:`null` to detect a dangling reference and avoid dereferencing
+it.
+
+The strategy of checking for :ada:`null` before trying to dereferencing the
+access value works fine for :ada:`I_1`. However, this strategy doesn't work for
+:ada:`I_2`: the access value is not :ada:`null` because it still points to an
+object that existed at some point. Therefore, the application will try to
+dereference :ada:`I_2`. The resulting behavior in this case is undefined.
+
+.. admonition:: Important
+
+    Dereferencing a dangling reference is erroneous: the behavior in this case
+    is undefined. Sometimes, the compiler is able to detect this situation and
+    introduce an :ref:`access check <Adv_Ada_Access_Check>`, which will fail
+    when trying to dereference the dangling reference, thereby raising the
+    :ada:`Constraint_Error` exception.
+
+    However, the compiler might not always be able to detect this, so the
+    behavior becomes unpredictable. The application might crash when
+    dereferencing the dangling reference, or it might just use some junk value
+    from the memory, giving the impression that everything is OK.
+
+    Because the behavior is erroneous and unpredictable, it should be
+    avoided at all costs! Again, it is the programmer's responsibility to be
+    very careful when using unchecked deallocation and avoid creating dangling
+    references.
+
+    .. admonition:: In the Ada Reference Manual
+
+        - `13.9.1 Data Validity <https://www.adaic.org/resources/add_content/standards/12rm/html/RM-13-9-1.html>`__
+        - `13.11.2 Unchecked Storage Deallocation <https://www.adaic.org/resources/add_content/standards/12rm/html/RM-13-11-2.html>`__
+
+Again, a possible way to fix the code example is to indicate that :ada:`I_2`
+doesn't refer to any object (when deallocating the object that :ada:`I_1`
+points to):
+
+.. code-block:: ada
+
+       Put_Line ("Freeing I_1");
+       I_2 := null;
+       Free (I_1);
+
+With this fix, :ada:`I_2` isn't a dangling reference anymore after the call to
+:ada:`Free (I_1)`.
```

Comment:
It won't be practical for the programmer to make every possible dangling reference become null if the calls to Free are strewn throughout the code. That means the suggested design is to not use Free in client code, but instead hide its use within the bigger abstraction.  That way all the occurrences of the calls to Free are in one package, and the programmer of that package then can prevent dangling references. For example, if we have a bank account that involves dynamic allocation for some part of an account, the bank account is a private type (or limited), and the access values are hidden in the private part. Thus when an account is closed the allocated memory can be freed without the possibility of leaving dangling references. That will work as long as the reference cannot be copied when copying bank account objects, in which case we'd want the type to be limited as well as private. Or we'd make it a controlled type so that Finalize can call Free. I don't know if you will want to mention this design idiom, and I'm not sure  a bank account is a good example! :-)

Reply:
Thanks a lot for this comment! I agree, we should describe good design patterns for access types in a separate section. (I added a todo item for that, it's on my list now.)